### PR TITLE
Vendor interfaces

### PR DESCRIPTION
**Pull request by fmauch**
This MR adds an overview over a couple of vendor interfaces. This should serve as a guide to develop an industrial-like robot interface ontop of existing vendor interfaces. The idea is to pickup the suggestion from https://github.com/ros-industrial/rep/blob/master/rep-I0003.rst

This consists of the following sections:
 - [x] Introduction - why do we do this?
 - [x] Overview of existing vendor interfaces
 - [x] Summary - What are common interfaces in many robots and what could be a resulting interface?

### CODE REVIEW

File: source/industry/Fanuc.rst [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+.. _Karel: https://www.tristarcnc.com/News/KarelProgrammingLanguage
+
+Cartesian trajectories for Fanuc robots
+
+  =================================   =======================================
+  Teach pendant                       FANUC iPendant touch
+  Programming / simulation software   ROBOGUIDE
+  Software                            
+  User interface                      Fanuc iHMI (Intelligent Human Machine Interface)
+  Programming language                FANUC `Karel`_ (derived from Pascal)
```
Comment (gavanderhoorn) [inline source/industry/Fanuc.rst:17]: Karel is a programming language on Fanucs, but it's not used to control any motion.

Much of the content in this page seems to focus on system variables which do influence motion, but are typically not directly changed by user code.

TP is the programming language which is used for motion programs. It almost never directly changes the system variables mentioned in the *Specification of velocity* section (and further down).

File: source/industry/Fanuc.rst [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+* **$USEMAXACCEL**: enables 'fast acceleration' feature
+
+
+Blending
+~~~~~~~~
+
+Taught positions can either be fly-by points, or stop points:
+
+* **FINE**: motion stops robot arm briefly at each way point
+* **CNT**: robot approaches to the point with a distance specified by the CNT value without ever actually reaching the point.
```
Comment (gavanderhoorn) [inline source/industry/Fanuc.rst]: This list is missing `CR` (Corner Radius), which can be used to (more) precisely define *when* (and where) blending should be started.

File: source/industry/Fanuc.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+Dynamic Path Modifier (DPM)
+
+* dynamic path modification using sensor data, so robot's path can be adapted in real-time
+* an external sensor provides postion and orientation offset for the next destination
+* applicable to multiple groups
+* possible applications:
+
+    * Weave operations
+    * Stationary tracking
+    * Orientation control
```
Comment (gavanderhoorn) [inline source/industry/Fanuc.rst:109]: The `J519` option (or *Stream Motion*) is missing here as well as `R912` (or *Remote Motion Interface*).

The former is similar to KUKA RSI and ABB EGM, the latter is essentially a drip-feed for TP programs.

Unfortunately `J519` availability is being restricted to the US only. `R912` is available world-wide.

**Review by gavanderhoorn — COMMENTED** [review]
Some comments about the Fanuc page.

File: source/industry/Fanuc.rst (hunk 2 above)
Comment (LeaStef) [reply source/industry/Fanuc.rst]: I aggree, CR seems to be a more advanced option but should be included here.
```suggestion
* **CNT** (continuous): robot approaches to the point with a distance specified by the CNT value without ever actually reaching the point, so the robots arm moves in a continuous trajectory
* **CR** (corner radius): like CNT, but specifying a radius for corner rounding allows to precisely define the shape of the blended motion
```

File: source/industry/Fanuc.rst (hunk 3 above)
Comment (LeaStef) [reply source/industry/Fanuc.rst:109]: That is very interesting. However, I could not find much information about R912.
```suggestion
    * Orientation control
    
J519  (Stream Motion)     
    * external protocol for:
        * path trajectory planning
        * near-real time streaming of the path trajectory to the robot 
        * enabling highly flexible and dynamic applications
    
R912 (Remote Motion Interface)       
    * drip-feed for TP programs
    
```

Comment (gavanderhoorn) [reply source/industry/Fanuc.rst:109]: >  I could not find much information about R912.

It's a relatively new option. That's probably the reason.

Note that the restriction to the US of J519 is only something I've heard about. It's not official policy necessarily.

Comment (LeaStef) [reply source/industry/Fanuc.rst:109]: ah thanks. I therefor decided to neglect the information about the availability

File: source/industry/Fanuc.rst (hunk 1 above)
Comment (fmauch) [reply source/industry/Fanuc.rst:17]: @stefanscherzinger @LeaStef Would one of you please work that in?

File: source/industry/Fanuc.rst [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+.. _Karel: https://www.tristarcnc.com/News/KarelProgrammingLanguage
+
+Cartesian trajectories for Fanuc robots
+
+  =================================   =======================================
+  Teach pendant                       FANUC iPendant touch
+  Programming / simulation software   ROBOGUIDE
+  Software                            
+  User interface                      Fanuc iHMI (Intelligent Human Machine Interface)
+  Programming language                FANUC `Karel`_ (derived from Pascal)
```
Comment (LeaStef) [inline source/industry/Fanuc.rst:17]:
```suggestion
  Programming language                TP (Teach Pendant), Karel
```

File: source/industry/Fanuc.rst (hunk 1 above)
Comment (fmauch) [reply source/industry/Fanuc.rst:17]:
```suggestion
  Programming language                TP (Teach Pendant), Karel
```

@LeaStef I put your suggestion here to keep it in this discussion thread.

@gavanderhoorn Would hit your point raised? I'm not sure whether you are also stating that the content below might not reflect the relevant aspects.

Comment (gavanderhoorn) [reply source/industry/Fanuc.rst:17]: The system variables mentioned in this article are "never" directly manipulated from/in TP programs.

They encode system configuration, which is manipulated through menus, keys on the TP and/or statements in a TP program, but never directly (ie: you don't really write `$MCR[].$GENOVERRIDE = 100` in TP, but `OVERRIDE=100%`).

I'm not saying those system variables must not be mentioned, I'm asking (implicitly) why they *are* mentioned. As, as I wrote, user programs don't use them to configure motion. Nor would any driver I believe (as motion *must* be initiated/controlled from TP, and so drivers would also use TP for that).

File: source/industry/Fanuc.rst [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    * `$MCR.$GENOVERRIDE`: general override
+    * `$MCR_GRP.$PROGOVERRIDE`: program override
+
+* Manual Motion Speed
+
+    * Joint Speed (in joint units) equals: `$PARAM_GROUP[i].$SPEEDLIMJNT` * `$MCR[].$GENOVERRIDE` / 100  * `$SCR_GRP[i].$JOGLIM_JNT` / 100
+    * Cartesian Translational Speed (in mm/sec) equals: $PARAM_GROUP[i].SPEEDLIM * $MCR[].$GENOVERRIDE / 100 * $SCR_GRP[i].$JOGLIM / 100
+    * Cartesian Rotational Speed (in mm/sec) equals: ($PARAM_GROUP[i].ROTSPEEDLIM * $MCR[].$GENOVERRIDE / 100 * $SCR_GRP[i].$JOGLIMROT / 100
+
+* Programmed Motion Speed
```
Comment (LeaStef) [inline source/industry/Fanuc.rst]:
```suggestion
As motions are initiated and controlled in TP the user can only adapt the robot's motion speed with TP. System configurations and overrides influence the velocity additionally.
```

File: source/industry/Fanuc.rst (hunk 1 above)
Comment (LeaStef) [reply source/industry/Fanuc.rst:17]: Sorry, for taking s long. After all I agree and adapted it accordingly.

File: source/industry/Fanuc.rst (hunk 5 above)
Comment (fmauch) [reply source/industry/Fanuc.rst]: The last word is misspelled in the suggestion above.

File: source/industry/UR.rst [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+   
+    * For *MoveJ* in deg/s:  **maximum joint speed**
+    * For *MoveL* in mm/s:  **desired tool speed**
+      
+Global specification of velocity are done via two safety-related functions:
+      
+     ===========  ===========================     ===========
+     Function     Description                     Limit
+     ===========  ===========================     ===========
+     Joint speed  Max. angular joint speed        30 ◦/s
```
Comment (LeaStef) [inline source/industry/UR.rst]:
```suggestion
     Joint speed  Max. angular joint speed        180 ◦/s
```
exception is the wrist3 of UR3s with 360°/s, but I would not mention it as it is to specific.

File: source/industry/UR.rst [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+     TCP speed    Max. speed of the robot TCP     250 mm/s
```
Comment (LeaStef) [inline source/industry/UR.rst]:
```suggestion
     TCP speed    Max. speed of the robot TCP     5000 mm/s
```

File: source/industry/UR.rst [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+Specification of velocity
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The robot's speed is defined in form of an argument by the move command.
+   
+    * For *MoveJ* in deg/s:  **maximum joint speed**
+    * For *MoveL* in mm/s:  **desired tool speed**
+      
+Global specification of velocity are done via two safety-related functions:
```
Comment (LeaStef) [inline source/industry/UR.rst]:
```suggestion
Global specification of velocity are done sparately for joint and TCP speed. The limits, which depend on the robot version, are stated in the table below. The actual speed limits also depend on the robot configuration.
```

File: source/industry/summary.rst [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+-----------------------------
+
+
++--------------+------------+--------------+------------+-----------+
+|              | **linear** | **circular** | **spline** | **joint** |
+|              | Cartesian  | Cartesian    | Cartesian  |           |
++==============+============+==============+============+===========+
+| KUKA         | yes        | yes          | yes        | yes       |
++--------------+------------+--------------+------------+-----------+
+| UR           | yes        | yes          | yes        | yes       |
```
Comment (LeaStef) [inline source/industry/summary.rst]:
```suggestion
| UR           | yes        | yes          | no         | yes       |
```
UR supports blends, but is is not the same as splines.

File: source/industry/Fanuc.rst (hunk 5 above)
Comment (LeaStef) [reply source/industry/Fanuc.rst]: fixed the typo

File: source/industry/UR.rst (hunk 6 above)
Comment (fmauch) [reply source/industry/UR.rst]: It might be good to drop that into a comment or footnote below the table.

File: source/industry/UR.rst [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+Specification of velocity
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The robot's speed is defined in form of an argument by the move command.
+   
+    * For *MoveJ* in deg/s:  **maximum joint speed**
+    * For *MoveL* in mm/s:  **desired tool speed**
+      
+Global specification of velocity are done sparately for joint and TCP speed. The limits, which depend on the robot version, are stated in the table below. The actual speed limits also depend on the robot configuration.
```
Comment (LeaStef) [inline source/industry/UR.rst]:
```suggestion
Global specification of velocity are done separately for joint and TCP speed. The limits, which depend on the robot version, are stated in the table below [#]_ . The actual speed limits also depend on the robot configuration.
```

File: source/industry/UR.rst [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+      
+     ===========  ===========================     ===========
+     Function     Description                     Limit
+     ===========  ===========================     ===========
+     Joint speed  Max. angular joint speed        30 ◦/s
+     TCP speed    Max. speed of the robot TCP     5000 mm/s
+     ===========  ===========================     ===========
+
```
Comment (LeaStef) [inline source/industry/UR.rst:118]:
```suggestion
.. [#] An exception is the wrist3 of UR3s with max. angular speed of 360°/s.
```

File: source/industry/UR.rst (hunk 6 above)
Comment (LeaStef) [reply source/industry/UR.rst]: done

File: source/industry/UR.rst (hunk 11 above)
Comment (fmauch) [reply source/industry/UR.rst:118]: I think, I would prefer, if you could either edit your original suggestion or make a commit. This little suggestion bits make it hard to follow up.

If you make a commit, then you can also see the output locally and we get a more comprehensive state. For example, with the change suggested here, I cannot see that defining the footnote actually results in a warning that the footnote isn't referenced.

File: source/industry/UR.rst (hunk 6 above)
Comment (fmauch) [reply source/industry/UR.rst]: Fixed in bfc6a8f

File: source/industry/UR.rst (hunk 11 above)
Comment (fmauch) [reply source/industry/UR.rst:118]: Fixed in bfc6a8f
